### ISSUE TITLE
Ideas for RETURNN behaviour changes

### ISSUE THREAD

**Issue by JackTemaki** [rwth-i6/returnn#508]
As there are now many different ways (as in parameter names, syntax styles etc...) to make RETURNN do the same things, it might be a good idea to add a `behaviour_version` parameter to prohibit using "deprecated" parameters and suboptimal behaviour of RETURNN. It can also be used to change the default behaviour of some layers which is currently suboptimal, such as the batch norm layers. With each desired change (e.g. commit or PR) to the behaviour, the version number would have to be increased by one.

Here are some things we already recommend to avoid according to the documentation, e.g.:
 - `num_inputs` and `num_outputs` should not be used in favor of `extern_data`
 - using the "dict" syntax for `extern_data` instead of the tuple syntax
 - prohibit using old theano parameters such as "multiprocessing" when working with tensorflow
 - prohibit direct optimizer parameters, so writing `optimizer = {'class': "adam"}` or maybe `optimizer="adam"` instead of `adam=True`
 - always use explicit input and target definitions (no default fallback to "data" or "classes")
 
 Then there are other redundant parameters which I can not name explicitely, but I think we have:
  -  overlapping/redundant parameters names for the dataset (e.g. `eval` versus `search_data` or so)
  - learning_rate vs learning_rates
 
 And then we have sub-optimal default parameters, that should be changed or not have a default in the first place, e.g.:
 - batch norm layer parameters
 - `ExtractAudioFeatures` should have no defaults, so that the feature settings are clearly visible in the config
 - no default `lstm` unit for rec layer
 
Other things I would personally recommend doing in configs, e.g.:
 - always provide an explicit `available_for_inference` for the `extern_data` entries
 - always use `seq_order_control_dataset` within a `MetaDataset`
 - not using `out_type` for any layer except maybe eval layer. Although here I am not sure if this has any major implications. Maybe there are some networks with rather instable construction and you want to have this fixed by hand.
 
There is probably a lot more which needs to be collected/discussed, so this is just intended to give some initial ideas.

**Comment by albertz** [comment]
As discussed, this is just an example list. For every individual potential change, we should make a separate issue, and discuss things there. Every such issue should use the potential-new-behavior tag (and maybe reference this issue here).

This issue here probably can be closed once we have a first version of `behaviour_version`.
The setup of `behaviour_version` can probably be done in a similar way as `use_tensorflow`.
(Btw, `use_tensorflow` is also a default we can change by this.)

**Comment by JackTemaki** [comment]
I would suggest then to start with restricting global (root level) parameters for the "initial" PR introducing `behaviour_version=1`, I will open a new explicit issue for this later on .

**Comment by JackTemaki** [comment]
For  `behaviour_version=1`, I would recommend to start with global parameters.

__Parameters that are no longer allowed to be used__  (RETURNN will throw an error) __:__

Agreed on:
 
Discussion:
  - disallowing `num_inputs` and `num_outputs` #513
  - All boolean optimizer parameters, such as: `adam`, `nadam`... etc in favour of using `optimizer` #514 
  - `learning_rate` vs `learning_rates`? #515 
  - `batching` is redundant with `seq_ordering` in the datasets, and is a misleading name #516 
  - Theano parameters? 
 
 
 Here is a full list of all possible parameters:
 https://gist.github.com/JackTemaki/b8c23b8eba61e4951ea22561f30e54a1
 
Many of those only belong to the Theano backend, and can possibly be ignored

**Comment by albertz** [comment]
> For behaviour_version=1, I would recommend to start with global parameters.
> Discussion ...

No, I would not wait until we have some sort of discussion finished here (which might take weeks, or never).

I would just start by one single change (e.g. `num_inputs`/`num_outputs` + disallowing tuples) and introduce `behaviour_version=1` for that.

Also, I would prefer to keep each increment of `behaviour_version` mostly for a single logical change, and not mix up different things (eg `optimizer` and `num_inputs` are totally independent of each other, should not be mixed up).

Also, you do not need to specifically list and collect each thing here in a comment because GitHub will automatically do that.

**Comment by JackTemaki** [comment]
> Also, you do not need to specifically list and collect each thing here in a comment because GitHub will automatically do that.
If we take each issue as one change this is not needed, that is correct

**Comment by albertz** [comment]
Btw, as this came up now with a proposed change on enforcing `available_for_inference`, but I want to put this more general:

In principle, I would say that it should be easy for most users to use their existing setups and increase `behaviour_version`, and if they have followed good practice and our general recommendations, it would all still work (maybe just get better, by better random init defaults or so).

I'm not sure on changes which would break the majority of setups. E.g. enforcing `available_for_inference` is such an example. I think really no-one explicitly specifies this (unless really needed, when the default is wrong). (Let's not discuss `available_for_inference` specifically here but leave that to a separate issue.)

So for a user which wants to keep up by using always the latest `behaviour_version`, how simple should that be for him/her? Should we still try to keep this simple, and avoid breaking changes unless for very good reasons? But maybe this just has to be discussed and decided for each change individually, and there is no generic answer.

**Comment by albertz** [comment]
Ah, another thing: We should have a dedicated documentation for this. Where users can read what is being changed for each `behavior_version`. (Maybe similar as the [changelog](https://github.com/rwth-i6/returnn/blob/master/CHANGELOG.md).)

**Comment by albertz** [comment]
Btw, one other idea, maybe should be discussed now before we start working on any of this:

Instead of having `behavior_version` (an integer), maybe we should explicitly (or additionally) have flags for each change?

I wonder for the case of a Sisyphus setup which consists of many experiments, where maybe some older experiments use some older `behavior_version`, and some others use a newer `behavior_version`. And when Sisyphus interacts with these configs, some interactions might then also depend on the `behavior_version` (e.g. random example: the search job needs to use `search_data`, and sth changed there). But then this might clash. Sisyphus is just an example. There might be other cases maybe (not sure how often, or how difficult it is to deal with that). If we also have explicit flags, we can maybe overwrite some of this, like `behavior_enable = {flag_x}` or so.

Just an idea. Maybe can also be added later.

**Comment by JackTemaki** [comment]
> And when Sisyphus interacts with these configs, some interactions might then also depend on the behavior_version

This would only affect things hard-coded inside RETURNN jobs, which should be reduced to the absolute minimum anyway. Having the integer is absolutely fine, because you can just parse this from the input config. We do not want to allow any mixture of behaviours, this will end up in total chaos.

**Comment by albertz** [comment]
> > And when Sisyphus interacts with these configs, some interactions might then also depend on the behavior_version
> 
> This would only affect things hard-coded inside RETURNN jobs, which should be reduced to the absolute minimum anyway.

But there is still some (minimum) needed interaction. And the behavior of some of this could potentially maybe change with a new `behavior_version`. Or we say we never want such changes.

But maybe I'm overthinking this now, and practically we will never get to this problem.

**Comment by JackTemaki** [comment]
> Or we say we never want such changes.

The changes are allowed, and can easily be dealt with in Sisyphus. Different `behaviour_version` means a different hash, so if the internal logic changes then it is completely acceptable.

**Comment by albertz** [comment]
> > Or we say we never want such changes.
> 
> The changes are allowed, and can easily be dealt with in Sisyphus. Different `behaviour_version` means a different hash, so if the internal logic changes then it is completely acceptable.

This is not what I meant. I meant that e.g. some job, e.g. search job, might require some specific behavior (some specific min `behaviour_version`).

**Comment by JackTemaki** [comment]
> This is not what I meant. I meant that e.g. some job, e.g. search job, might require some specific behavior

Ah okay, no I do not think this will ever become a problem. And if for some reason this should ever happen, the job can just get an assert that the behaviour_version needs to be higher and thats it.

**Comment by albertz** [comment]
Another aspect: Sometimes (or even often), the discussion will not be about behavior change, but just about removing some functionality, in favor of some better way.

We don't strictly need to disallow this then. We can just put a deprecation warning with information what to use instead.

The deprecation warning should probably be printed in any case, no matter the `behavior_version`. So the question is just, should sth strictly be disallowed with a new `behavior_version` or is the deprecation warning enough?

Maybe this again has to be decided just individually per case.

**Comment by JackTemaki** [comment]
I would prefer to strictly disallow this with new behavior versions, because warnings are often missed.

This is of course no reason to not add warnings to everything "deprecated" even before adding something as behavior version.

I think this is what we discuss in each issue: should we put only a warning, or should this be added as new behavior version and then be strictly enforced.

**Comment by JackTemaki** [comment]
In which file should we log the `behavior_version` changes? `VERSION_HISTORY.md` ? or together with other changes in the regular changelog?

**Comment by albertz** [comment]
> In which file should we log the `behavior_version` changes? `VERSION_HISTORY.md` ? or together with other changes in the regular changelog?

I would put it into the documentation directly. And I would specifically refer to `behavior_version`, to not leave any ambiguity what "version" refers to. So maybe `docs/behavior_version.rst` or so. And also reference it somewhere.

**Comment by albertz** [comment]
A comment on explicit behavior flags, or in general about a new behavior version: There are cases

* where we maybe do not want to enforce the user to adopt the new behavior, because the old behavior is still also fine, and this can be a choice by the user. E.g. explicit dim tags (#597) or unique dim tags (#632) are maybe features which can be optionally used.
* where the transition to the new behavior is non-trivial. For most of the currently proposed new behaviors, it is quite trivial for the user to adopt. E.g. explicitly define `extern_data`. Or always use `from` explicitly. Etc. But there are maybe other cases where the transition would not always be so trivial (e.g. #632).
* where it is not even clear yet whether the new behavior is better than the old behavior. Or the new behavior is still experimental.

In general, we would always recommend users to use the latest behavior version. But users might not follow this advice if it is too much effort. So maybe we only want to introduce a new behavior version if the transition is simple/trivial for the user. And otherwise behavior flags.

I'm not sure. This is maybe also as a transition phases, as long some behavior is still somewhat experimental or unclear (e.g. #597, #632), it can be a flag, and later it can be introduced as a new behavior version.

**Comment by JackTemaki** [comment]
> So maybe we only want to introduce a new behavior version if the transition is simple/trivial for the user. And otherwise behavior flags.

For experimental behavior we should definitely not update the `behavior_version`. Having extra flags is fine.
We should try to make any changes as clear and understandable is as possible, so only if new features are tested, documented and easy-to-understand examples exists we should adopt this into a new `behavior_version`.

**Comment by albertz** [comment]
I think all the open questions here have been clarified, and we are already having the first couple of behavior versions. So we can close this.

**Comment by albertz** [comment]
PyTorch will probably introduce sth similar (specifically for their param init defaults): https://github.com/pytorch/pytorch/pull/41638
```
torch.nn.init.set_init_version('1.7.1')
```
